### ISSUE TITLE
How should I use NKeys?

### ISSUE THREAD

**Issue by etrochim** [nats-io/nats.c#369]
I'm investigating the use of NKeys to perform authentication between nats.c clients and the NATS server. I'm only looking for something that is more secure than the simple username/password authentication that I'm using now. I don't believe I'm interested in a full decentralized JWT authentication setup.

I want to perform the authentication described in https://docs.nats.io/developing-with-nats/security/nkey. What isn't clear to me is exactly what should be done in the natsSignatureHandler callback. NATS client libraries for other languages each provide a library for working with NKeys directly. For instance, python has an nkey module that can be initialized with a seed value and sign data with it. Ideally, there would be a nats.c equivalent of that module which I could use to sign the provided nonce after securing getting the seed. But no such public API appears to exist.

Digging around I found natsKeys_Sign which is used by the JWT specific functions natsOptions_SetUserCredentialsCallbacks and natsOptions_SetUserCredentailsFromFiles. That would be perfect for this but is unfortunately inaccessible.

I could call libsodium directly to sign the nonce just like natsKeys_Sign does but it doesn't seem like that should be necessary given the design of the other libraries.

I feel like I'm misunderstanding how natsOptions_SetNKey is meant to be used. Could I have guidance on how to properly use that function? And was there a deliberate design decision to exclude an NKey API from the public interface and, if so, what is the intended alternative?

Thanks

**Comment by kozlovic** [comment]
@etrochim You are absolutely correct. I have added an option to do the signature automatically by being provide a file that contains the seed. The library takes care of clearing the memory where the seed is copied.
I have also added `nats_Sign()` to perform signature. Without this one, the option `natsOptions_SetNKey()` is a bit useless for users that don't have access to crypto libraries.

I will likely do a release early next week or sooner.
